### ISSUE TITLE
Autograd bug in mxnet-cu80: 0.12

### ISSUE THREAD

**Issue by roggiezhang-nv** [apache/incubator-mxnet#8494]
Note: Providing complete information in the most concise form is the best way to get help. This issue template serves as the checklist for essential information to most of the technical issues.

If the issue is non-technical, feel free to present the information in what you believe is the best form.

## Description
Autograd behaves not properly.

## Environment info (Required)
OS: Ubuntu 16.04
Package used (Python/R/Scala/Julia): Python
(I'm using ...)

## Error Message:
(Paste the complete error message, including stack trace.)

## Minimum reproducible example
(If you are using your own code, please provide a short script that reproduces the error. Otherwise, please provide link to the existing example.)

```
Snippet:

import mxnet as mx

a = mx.random.normal(shape = (3, 3, 3))
a.attach_grad()

b = 0
with mx.autograd.record():
    b = a

b.backward()
print(a.grad)

output: [[[ 0.  0.  0.]
  [ 0.  0.  0.]
  [ 0.  0.  0.]]

 [[ 0.  0.  0.]
  [ 0.  0.  0.]
  [ 0.  0.  0.]]

 [[ 0.  0.  0.]
  [ 0.  0.  0.]
  [ 0.  0.  0.]]]

<NDArray 3x3x3 @cpu(0)>

Expected: 
a.grad should equal to all one.
```

## Steps to reproduce
(Paste the commands you ran that produced the error.)

1. Run the above snippet.
2.

## What have you tried to solve it?

1. Assign b = a + 1
2.

**Comment by szha** [comment]
Your code just made `b` an alias of `a` which isn't computation.

The following three examples show
1. the "what have you tried to solve it"
2. the snippet you posted with the addition of id check for `a` and `b`. it printed `True`.
3. the snippet where in-place assignment happens. the same id check printed `False`

```python
In [1]: import mxnet as mx
   ...:
   ...: a = mx.random.normal(shape = (3, 3, 3))
   ...: a.attach_grad()
   ...:
   ...: b = 0
   ...: with mx.autograd.record():
   ...:     b = a+1
   ...: b.backward()
   ...: a.grad
   ...:
Out[1]:

[[[ 1.  1.  1.]
  [ 1.  1.  1.]
  [ 1.  1.  1.]]

 [[ 1.  1.  1.]
  [ 1.  1.  1.]
  [ 1.  1.  1.]]

 [[ 1.  1.  1.]
  [ 1.  1.  1.]
  [ 1.  1.  1.]]]
<NDArray 3x3x3 @cpu(0)>

In [2]: import mxnet as mx
   ...:
   ...: a = mx.random.normal(shape = (3, 3, 3))
   ...: a.attach_grad()
   ...:
   ...: b = 0
   ...: with mx.autograd.record():
   ...:     b = a
   ...: print(id(a) == id(b))
   ...: b.backward()
   ...: a.grad
   ...:
True
Out[2]:

[[[ 0.  0.  0.]
  [ 0.  0.  0.]
  [ 0.  0.  0.]]

 [[ 0.  0.  0.]
  [ 0.  0.  0.]
  [ 0.  0.  0.]]

 [[ 0.  0.  0.]
  [ 0.  0.  0.]
  [ 0.  0.  0.]]]
<NDArray 3x3x3 @cpu(0)>

In [3]: import mxnet as mx
   ...:
   ...: a = mx.random.normal(shape = (3, 3, 3))
   ...: a.attach_grad()
   ...:
   ...: b = mx.nd.zeros_like(a)
   ...: with mx.autograd.record():
   ...:     b[:] = a
   ...: print(id(a) == id(b))
   ...: b.backward()
   ...: a.grad
   ...:
False
Out[3]:

[[[ 1.  1.  1.]
  [ 1.  1.  1.]
  [ 1.  1.  1.]]

 [[ 1.  1.  1.]
  [ 1.  1.  1.]
  [ 1.  1.  1.]]

 [[ 1.  1.  1.]
  [ 1.  1.  1.]
  [ 1.  1.  1.]]]
<NDArray 3x3x3 @cpu(0)>
```

**Comment by roggiezhang-nv** [comment]
The question is 0.11 will do that computation.

**Comment by szha** [comment]
That was probably unintended. In fact, in 0.11.0 you don't even have to put computation in the autograd recording scope to get the same effect.

```
In [1]: import mxnet as mx

In [2]: mx.__version__
Out[2]: '0.11.0'

In [3]: import mxnet as mx
   ...:
   ...: a = mx.random.normal(shape = (3, 3, 3))
   ...: a.attach_grad()
   ...:
   ...: b = 0
   ...: with mx.autograd.record():
   ...:     b = a
   ...: print(id(a) == id(b))
   ...: b.backward()
   ...: a.grad
   ...:
True
Out[3]:

[[[ 1.  1.  1.]
  [ 1.  1.  1.]
  [ 1.  1.  1.]]

 [[ 1.  1.  1.]
  [ 1.  1.  1.]
  [ 1.  1.  1.]]

 [[ 1.  1.  1.]
  [ 1.  1.  1.]
  [ 1.  1.  1.]]]
<NDArray 3x3x3 @cpu(0)>

In [4]: import mxnet as mx
   ...:
   ...: a = mx.random.normal(shape = (3, 3, 3))
   ...: a.attach_grad()
   ...:
   ...: b = 0
   ...: b = a
   ...: print(id(a) == id(b))
   ...: b.backward()
   ...: a.grad
   ...:
True
Out[4]:

[[[ 1.  1.  1.]
  [ 1.  1.  1.]
  [ 1.  1.  1.]]

 [[ 1.  1.  1.]
  [ 1.  1.  1.]
  [ 1.  1.  1.]]

 [[ 1.  1.  1.]
  [ 1.  1.  1.]
  [ 1.  1.  1.]]]
<NDArray 3x3x3 @cpu(0)>

In [5]: import mxnet as mx
   ...:
   ...: a = mx.random.normal(shape = (3, 3, 3))
   ...: a.attach_grad()
   ...:
   ...: print(id(a) == id(b))
   ...: a.backward()
   ...: a.grad
   ...:
False
Out[5]:

[[[ 1.  1.  1.]
  [ 1.  1.  1.]
  [ 1.  1.  1.]]

 [[ 1.  1.  1.]
  [ 1.  1.  1.]
  [ 1.  1.  1.]]

 [[ 1.  1.  1.]
  [ 1.  1.  1.]
  [ 1.  1.  1.]]]
<NDArray 3x3x3 @cpu(0)>
```

**Comment by roggiezhang-nv** [comment]
OK. Does that more make sense if it returns with all ones from a user point. 

By the way, this is not the true problem I want to emphasize. I just found 0.12 behaves a lot differently in autograd as 0.11. I have a piece of code that has been broken in 0.12 while it's working in 0.11, here is the snippet of what it's doing:

```
a is a mx.nd.array
a.attach_grad()
loss = 0
with autograd.record():
    for each target:
          loss = loss + LOSS_FUNCTION(a, target)

loss.backward()
Update gradient with a.grad
```
In 0.11, this works without any issues. But after switching to 0.12, this is broken with a.grad equals to zero. 

So I just want to understand any important changes in 0.12 about autograd.

**Comment by roggiezhang-nv** [comment]
To narrow down the issue, I tried to install several beta versions of 0.11.1. And finally, I found the changelist breaks my code is between mxnet-cu80 0.11.1b20170829 and mxnet-cu80 0.11.1b20170901. Anybody could help on this? If you need, I'll list my code for you to check. 

Thanks.

**Comment by roggiezhang-nv** [comment]
Here is the code that behaves differently between those two builds: https://github.com/roggiezhang-nv/mxnet-sample.

You can run the notebook under notebooks/classic_style_transfer.ipynb to check that if the style transfer works with the two builds.

**Comment by roggiezhang** [comment]
Anybody could help here?

**Comment by szha** [comment]
I haven't had the time to look into the details of the code yet. I found [the part that your pseudo-code corresponds to](https://github.com/roggiezhang-nv/mxnet-sample/blob/master/python/ml/vision/style_transfer/classic.py#L103-L111), though I haven't found any apparent problem. How are you adjusting the weights given that your code doesn't use trainer?

BTW you may want to check out @zhanghang1989's work on [style-transfer](https://github.com/zhanghang1989/MXNet-Gluon-Style-Transfer).

**Comment by roggiezhang** [comment]
It does not need to adjust the weights of vgg since I only used that for feature extraction. The code logic is OK since I already made it work for 0.12, but not for 0.11.3. That's why I started this thread. I cannot find anything useful from source code of python api. And I debugged and found the difference after executing graph [https://github.com/apache/incubator-mxnet/blob/396943e22661f03867d103d134416541e7e4f2bb/python/mxnet/gluon/block.py#L394](line). For 0.12, the grad that I attached disappears after this line, which is not for 0.11.3.

By the way, here is the rough logic of my code:

1. Build pre-trained network vgg.
2. Get inputs and outputs symbols that are interested.
3. Use those inputs and outputs to build SymbolBlock.
4. Open autograd.record(), attach the grad for input, and feed the input to the SymbolBlock created above. Lastly, compute the loss and do grad. 

In theory, I should get a correct grad for input (that's how it works in 0.11.3), but it's broken in 0.12. Since it's related to native code, i'd have to ask for you guys to help.

**Comment by zhanghang1989** [comment]
Put `loss.backward()` inside the scope of `autograd.record()`
https://github.com/zhanghang1989/MXNet-Gluon-Style-Transfer/blob/master/main.py#L167-L176

**Comment by roggiezhang-nv** [comment]
I tried and that won't help.

**Comment by zhanghang1989** [comment]
Put all steps that requires grads in this scope including model forward

**Comment by roggiezhang-nv** [comment]
Interesting. This is solved since 0.12.1. Close this as fixed.
